pkg/scheduler/objects/node_iterator.go

```diff
@@ -19,103 +19,44 @@
+	// Reset the iterator to a clean state
 	Reset()
-}
-
-// All iterators extend the base iterator
-type baseIterator struct {
-	NodeIterator
-	countIdx int
-	size     int
-	nodes    []*Node
-}
-
-// Reset the iterator to start from the beginning
-func (bi *baseIterator) Reset() {
-	bi.countIdx = 0
-}
-
-// HasNext returns true if there is a next element in the array.
-// Returns false if there are no more elements or list is empty.
-func (bi *baseIterator) HasNext() bool {
-	return !(bi.countIdx+1 > bi.size)
-}
-
-// Next returns the next element and advances to next element in array.
-// Returns nil at the end of iteration.
-func (bi *baseIterator) Next() *Node {
-	if (bi.countIdx + 1) > bi.size {
-		return nil
-	}
 
-	value := bi.nodes[bi.countIdx]
-	bi.countIdx++
-	return value
+	// ForEachNode Calls the provided function on each Node object unil it returns false
+	ForEachNode(func(*Node) bool)
 }
 
-// Default iterator, wraps the base iterator.
-// Iterates over the list from the start, position zero, to end.
-type defaultNodeIterator struct {
-	baseIterator
+type treeIterator struct {
+	accept  func(*Node) bool
+	getTree func() *btree.BTree
+	tree    *btree.BTree
 }
 
-// Create a new default iterator
-func NewDefaultNodeIterator(schedulerNodes []*Node) NodeIterator {
-	it := &defaultNodeIterator{}
-	it.nodes = schedulerNodes
-	it.size = len(schedulerNodes)
-	return it
+func (ti *treeIterator) Reset() {
+	ti.tree = ti.getTree()
 }
 
-// Random iterator, wraps the base iterator
-// Iterates over the list from a random starting position in the list.
-// The iterator automatically wraps at the end of the list.
-type roundRobinNodeIterator struct {
-	baseIterator
-	startIdx int
-}
+func (ti *treeIterator) ForEachNode(f func(*Node) bool) {
+	ti.tree.Ascend(func(item btree.Item) bool {
+		node := item.(nodeRef).node
+		if ti.accept(node) {
+			return f(node)
+		}
 
-// The starting point is randomised in the slice.
-func NewRoundRobinNodeIterator(schedulerNodes []*Node) NodeIterator {
-	it := &roundRobinNodeIterator{}
-	it.nodes = schedulerNodes
-	it.size = len(schedulerNodes)
-	if it.size > 0 {
-		it.startIdx = rand.Intn(it.size)
-	}
-	return it
+		return true
+	})
 }
 
-// Next returns the next element and advances to next element in array.
-// Returns nil at the end of iteration.
-func (ri *roundRobinNodeIterator) Next() *Node {
-	// prevent panic on Next when slice is empty
-	if (ri.countIdx + 1) > ri.size {
-		return nil
-	}
-
-	// after reset initialize the rand seed based on number of nodes.
-	if ri.startIdx == -1 {
-		ri.startIdx = rand.Intn(ri.size)
+func NewTreeIterator(accept func(*Node) bool, getTree func() *btree.BTree) *treeIterator {
+	ti := &treeIterator{
+		getTree: getTree,
+		accept:  accept,
 	}
-
-	idx := (ri.countIdx + ri.startIdx) % ri.size
-	value := ri.nodes[idx]
-	ri.countIdx++
-	return value
-}
-
-func (ri *roundRobinNodeIterator) Reset() {
-	ri.countIdx = 0
-	ri.startIdx = -1
+	ti.tree = getTree()
```

Comment:
We can do an optimization here by not calling getTree() and storing the result. Instead, just do it in ForEachNode():

Instead of ti.tree.Ascend(...) use ti.getTree().Ascend(...).

This means we only compute the tree if ForEachNode is actually called.

Reply:
Alternatively, we could compute the tree here and simply iterate the same copy of it each time. We should not do both, however.